### PR TITLE
UserMedal improved and tests running well.

### PR DESCRIPTION
**Pull request by oricat8**

### CODE REVIEW

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java [hunk 1]
```diff
@@ -58,7 +57,11 @@
 
     @BeforeEach
     public void setUp() {
-        userMedalEntity = new UserMedalEntity("Walker", 1.0, 2.0, new ArrayList<>());
+        medal = new Medal();
+        medal.setDescription("You have to walk a lot!");
+        medal.setLevels(new ArrayList<>(Arrays.asList(5., 10., 25., 50., 100.)));
+        medal.setMedalIconPath(new byte[] {10, 55, 67, 89, 103, 116});
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java:63]: Codacy Issue found: ['10' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java [hunk 2]
```diff
@@ -58,7 +57,11 @@
 
     @BeforeEach
     public void setUp() {
-        userMedalEntity = new UserMedalEntity("Walker", 1.0, 2.0, new ArrayList<>());
+        medal = new Medal();
+        medal.setDescription("You have to walk a lot!");
+        medal.setLevels(new ArrayList<>(Arrays.asList(5., 10., 25., 50., 100.)));
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java:62]: Codacy Issue found: ['5.' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java [hunk 3]
```diff
@@ -8,16 +8,14 @@
...
 import org.pesmypetcare.webservice.error.DatabaseAccessException;
 import org.pesmypetcare.webservice.error.DocumentException;
 import org.pesmypetcare.webservice.thirdpartyservices.adapters.firestore.FirestoreCollection;
 import org.pesmypetcare.webservice.thirdpartyservices.adapters.firestore.FirestoreDocument;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java:18]: Codacy Issue found: [Using the '.*' form of import should be avoided - java.util.*.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

Comment (marclll) [inline src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java]: Hauries de indicar explicitament que importes i sols importar les classes necessaries

**Review by marclll — CHANGES_REQUESTED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java [hunk 4]
```diff
@@ -58,7 +61,11 @@
 
     @BeforeEach
     public void setUp() {
-        userMedalEntity = new UserMedalEntity("Walker", 1.0, 2.0, new ArrayList<>());
+        medal = new Medal();
+        medal.setDescription("You have to walk a lot!");
+        medal.setLevels(new ArrayList<>(Arrays.asList(5., 10., 25., 50., 100.)));
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java:66]: Codacy Issue found: ['5.' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java [hunk 5]
```diff
@@ -58,7 +61,11 @@
...
+        medal.setMedalIconPath(new byte[] {10, 55, 67, 89, 103, 116});
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoTest.java:67]: Codacy Issue found: ['10' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Review by marclll — DISMISSED** [review]
Ok

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

**Review by santidrj — COMMENTED** [review]
I have a couple of questions about the feature:
1. When you create a new medal does the same method create a UserMedal for all the users in the DB?
2. When you get a UserMedal, does that method get the necessary data from it's corresponding Medal and saves it inside the UserMedal you are getting?
3. There also should be a method to create all the necessary UserMedals for a new user, so that UserDao could call that method when creating a user.

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage decreased per file
===========================
- src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java  -13
- src/main/java/org/pesmypetcare/webservice/service/medalmanager/UserMedalServiceImpl.java  -25
- src/main/java/org/pesmypetcare/webservice/controller/medalmanager/UserMedalRestController.java  -20
         

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java [hunk 6]
```diff
@@ -82,4 +88,20 @@ private void initializeWithCollectionPath() {
         batch = dbCol.batch();
         path = Path.ofCollection(Collections.medals);
     }
+
+    /**
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:92]: Codacy Issue found: [First sentence should end with a period.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java [hunk 7]
```diff
@@ -70,6 +80,23 @@ public Object getField(String owner, String name, String field) throws DatabaseA
         return dbDoc.getDocumentField(medalPath, field);
     }
 
+    /**
+     * Create all user medals when user is created.
+     * @param username Username of the user.
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When the document does not exist
+     */
+    public void createAllUserMedals (String username) throws DatabaseAccessException, DocumentException {
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java:89]: Codacy Issue found: ['(' is preceded with whitespace.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java [hunk 8]
```diff
@@ -70,6 +80,23 @@ public Object getField(String owner, String name, String field) throws DatabaseA
...
+        path = Path.ofCollection(Collections.medals);
+        List<DocumentSnapshot> medalsDocuments = dbCol.listAllCollectionDocumentSnapshots(path);
+        for (DocumentSnapshot medalDocument : medalsDocuments) {
+            MedalEntity medal = medalDocument.toObject(MedalEntity.class);
+            UserMedalEntity userMedal = new UserMedalEntity(medal.getName(), 0., 0.,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java:94]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java [hunk 9]
```diff
@@ -82,4 +88,20 @@ private void initializeWithCollectionPath() {
...
+     * Insert the new medal to all users
+     * @param medal The new medal we had created before
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When the document does not exist
+     */
+    private void addMedalsToUsers(MedalEntity medal) throws DatabaseAccessException, DocumentException {
+        path = Path.ofCollection(Collections.used_usernames);
+        List<DocumentSnapshot> allUsers = dbCol.listAllCollectionDocumentSnapshots(path);
+        for (DocumentSnapshot user: allUsers) {
+            UserMedalEntity medalEntity = new UserMedalEntity(medal.getName(), 0., 0.,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:102]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java [hunk 10]
```diff
@@ -82,4 +88,21 @@ private void initializeWithCollectionPath() {
         batch = dbCol.batch();
         path = Path.ofCollection(Collections.medals);
     }
+
+    /**
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:92]: Codacy Issue found: [First sentence should end with a period.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java [hunk 11]
```diff
@@ -70,6 +80,24 @@ public Object getField(String owner, String name, String field) throws DatabaseA
...
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When the document does not exist
+     */
+    public void createAllUserMedals(String username) throws DatabaseAccessException, DocumentException {
+        path = Path.ofCollection(Collections.medals);
+        List<DocumentSnapshot> medalsDocuments = dbCol.listAllCollectionDocumentSnapshots(path);
+        UserMedalEntity userMedal;
+        for (DocumentSnapshot medalDocument : medalsDocuments) {
+            MedalEntity medal = medalDocument.toObject(MedalEntity.class);
+            userMedal = new UserMedalEntity(medal.getName(), 0., 0.,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java:95]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java [hunk 12]
```diff
@@ -82,4 +88,21 @@ private void initializeWithCollectionPath() {
...
+     * @param medal The new medal we had created before
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When the document does not exist
+     */
+    private void addMedalsToUsers(MedalEntity medal) throws DatabaseAccessException, DocumentException {
+        path = Path.ofCollection(Collections.used_usernames);
+        List<DocumentSnapshot> allUsers = dbCol.listAllCollectionDocumentSnapshots(path);
+        UserMedalEntity medalEntity;
+        for (DocumentSnapshot user: allUsers) {
+            medalEntity = new UserMedalEntity(medal.getName(), 0., 0.,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:103]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java [hunk 13]
```diff
@@ -82,4 +88,21 @@ private void initializeWithCollectionPath() {
...
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:103]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Review by santidrj — DISMISSED** [review]
:+1:

**Review by marclll — DISMISSED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java (hunk 11 above)
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java:95]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java [hunk 14]
```diff
@@ -82,4 +88,20 @@ private void initializeWithCollectionPath() {
...
+    /**
+     * Insert the new medal to all users.
+     * @param medal The new medal we had created before
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When the document does not exist
+     */
+    private void addMedalsToUsers(MedalEntity medal) throws DatabaseAccessException, DocumentException {
+        path = Path.ofCollection(Collections.used_usernames);
+        List<DocumentSnapshot> allUsers = dbCol.listAllCollectionDocumentSnapshots(path);
+        UserMedalEntity medalEntity  = new UserMedalEntity(medal.getName(), 0., 0.,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:101]: Codacy Issue found: [Use a single space to separate non-whitespace characters.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Review by marclll — DISMISSED** [review]

**Review by marclll — DISMISSED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

File: src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java (hunk 14 above)
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java:101]: Codacy Issue found: [Use a single space to separate non-whitespace characters.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage decreased per file
===========================
- src/main/java/org/pesmypetcare/webservice/service/medalmanager/UserMedalServiceImpl.java  -25
- src/main/java/org/pesmypetcare/webservice/dao/medalmanager/MedalDaoImpl.java  -23
- src/main/java/org/pesmypetcare/webservice/dao/medalmanager/UserMedalDaoImpl.java  -30
         

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/entity/medalmanager/UserMedalEntity.java  2
         

Complexity decreasing per file
==============================
+ src/main/java/org/pesmypetcare/webservice/entity/medalmanager/MedalEntity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5630992&bid=18442980)

**Review by marclll — APPROVED** [review]

**Review by santidrj — APPROVED** [review]
:+1: :trollface:

**Pull request merged**